// @ngxs/form-plugin/esm2015/src/actions.js
export class UpdateFormStatus {
    constructor(payload) {
        this.payload = payload;
    }
}
UpdateFormStatus.type = '[Forms] Update Form Status';
export class UpdateFormValue {
    constructor(payload) {
        this.payload = payload;
    }
}
UpdateFormValue.type = '[Forms] Update Form Value';
export class UpdateForm {
    constructor(payload) {
        this.payload = payload;
    }
}
UpdateForm.type = '[Forms] Update Form';
export class UpdateFormDirty {
    constructor(payload) {
        this.payload = payload;
    }
}
UpdateFormDirty.type = '[Forms] Update Form Dirty';
export class SetFormDirty {
    constructor(payload) {
        this.payload = payload;
    }
}
SetFormDirty.type = '[Forms] Set Form Dirty';
export class SetFormPristine {
    constructor(payload) {
        this.payload = payload;
    }
}
SetFormPristine.type = '[Forms] Set Form Pristine';
export class UpdateFormErrors {
    constructor(payload) {
        this.payload = payload;
    }
}
UpdateFormErrors.type = '[Forms] Update Form Errors';
export class SetFormDisabled {
    constructor(payload) {
        this.payload = payload;
    }
}
SetFormDisabled.type = '[Forms] Set Form Disabled';
export class SetFormEnabled {
    constructor(payload) {
        this.payload = payload;
    }
}
SetFormEnabled.type = '[Forms] Set Form Enabled';
