// @ngxs/form-plugin/esm5/src/actions.js
var UpdateFormStatus = /** @class */ (function () {
    function UpdateFormStatus(payload) {
        this.payload = payload;
    }
    return UpdateFormStatus;
}());
export { UpdateFormStatus };
UpdateFormStatus.type = '[Forms] Update Form Status';
var UpdateFormValue = /** @class */ (function () {
    function UpdateFormValue(payload) {
        this.payload = payload;
    }
    return UpdateFormValue;
}());
export { UpdateFormValue };
UpdateFormValue.type = '[Forms] Update Form Value';
var UpdateForm = /** @class */ (function () {
    function UpdateForm(payload) {
        this.payload = payload;
    }
    return UpdateForm;
}());
export { UpdateForm };
UpdateForm.type = '[Forms] Update Form';
var UpdateFormDirty = /** @class */ (function () {
    function UpdateFormDirty(payload) {
        this.payload = payload;
    }
    return UpdateFormDirty;
}());
export { UpdateFormDirty };
UpdateFormDirty.type = '[Forms] Update Form Dirty';
var SetFormDirty = /** @class */ (function () {
    function SetFormDirty(payload) {
        this.payload = payload;
    }
    return SetFormDirty;
}());
export { SetFormDirty };
SetFormDirty.type = '[Forms] Set Form Dirty';
var SetFormPristine = /** @class */ (function () {
    function SetFormPristine(payload) {
        this.payload = payload;
    }
    return SetFormPristine;
}());
export { SetFormPristine };
SetFormPristine.type = '[Forms] Set Form Pristine';
var UpdateFormErrors = /** @class */ (function () {
    function UpdateFormErrors(payload) {
        this.payload = payload;
    }
    return UpdateFormErrors;
}());
export { UpdateFormErrors };
UpdateFormErrors.type = '[Forms] Update Form Errors';
var SetFormDisabled = /** @class */ (function () {
    function SetFormDisabled(payload) {
        this.payload = payload;
    }
    return SetFormDisabled;
}());
export { SetFormDisabled };
SetFormDisabled.type = '[Forms] Set Form Disabled';
var SetFormEnabled = /** @class */ (function () {
    function SetFormEnabled(payload) {
        this.payload = payload;
    }
    return SetFormEnabled;
}());
export { SetFormEnabled };
SetFormEnabled.type = '[Forms] Set Form Enabled';
