// @angular-mdc/web/esm2015/elevation.js
/**
 * @license
 * Copyright (c) 2018 Dominic Carretto
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/trimox/angular-mdc-web/blob/master/LICENSE
 */
import { Directive, ElementRef, Input, NgModule } from '@angular/core';
import { toNumber } from '@angular-mdc/web/common';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class MdcElevation {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        this.elementRef = elementRef;
        this._mdcElevation = 0;
    }
    /**
     * @return {?}
     */
    get mdcElevation() { return this._mdcElevation; }
    /**
     * @param {?} value
     * @return {?}
     */
    set mdcElevation(value) {
        this._mdcElevation = toNumber(value);
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        const /** @type {?} */ change = changes['mdcElevation'];
        if (change.currentValue < 0 || change.currentValue > 24) {
            throw new Error(`Valid mdc-elevation values are 0 through 24`);
        }
        if (!change.isFirstChange()) {
            this.elementRef.nativeElement.classList.remove(`mdc-elevation--z${change.previousValue}`);
        }
        this.elementRef.nativeElement.classList.add(`mdc-elevation--z${change.currentValue}`);
    }
}
MdcElevation.decorators = [
    { type: Directive, args: [{
                selector: '[mdcElevation]',
                exportAs: 'mdcElevation'
            },] },
];
/** @nocollapse */
MdcElevation.ctorParameters = () => [
    { type: ElementRef }
];
MdcElevation.propDecorators = {
    mdcElevation: [{ type: Input }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class MdcElevationModule {
}
MdcElevationModule.decorators = [
    { type: NgModule, args: [{
                exports: [MdcElevation],
                declarations: [MdcElevation],
            },] },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

export { MdcElevationModule, MdcElevation };
